### DISCUSSION TITLE
A suggestion concerning a reflect-metadata package

### DISCUSSION THREAD

**Discussion by thephoenixofthevoid** [loopbackio/loopback-next#7570] [Ideas]
Since the code runs only in the modern environment (node 10 recently gets deprecated), all kinds of shims for standard ES6 classes packed deep inside of "reflect-metadata" make no sense.

Recently I remastered this package and come up with a very clear solution that has exactly the same behavior as the original one.
I would like to share it with a community: https://github.com/thephoenixofthevoid/reflect-metadata-ng
If you have any questions about it, ask here or create an issue in the above repo.

**Comment by raymondfeng** [comment]
Interesting. What `modern` features does your package depend on?

**Reply by thephoenixofthevoid** [reply]
I simply removed all shims (for Set, Map, WeakSet, WeakMap, ES6 Iterators for Set/Map) and all glue code that made the choice between shim and native implementation. The rest of the code is **identical**. It's a repack not a re-implementation

**Reply by thephoenixofthevoid** [reply]
WeakMapShim/WeakSetShim were naive: they leaked memory, so I think it's preferable to crash at the start and ask to update a node to a version 4, then leak memory in production. Considering that v10 is deprecated already, I do not think it's a problem at all. Strictly speaking, there is a chance of false-positive Shim activation -- triggering logic is very complex and cryptic. During the inspection of that logic's source code, I noticed that there is a way to extract valuable parts, and then wire them to globals directly

**Reply by thephoenixofthevoid** [reply]
I quickly generated the diff, it's here: https://github.com/thephoenixofthevoid/reflect-metadata-ng/blob/main/diff_original_ng.html

But Github doesn't set the correct MIME type and the one of the solutions is to use githack:
```
raw.githack.com serves raw files directly from various source
code hosting services with proper Content-Type headers 
```
So: https://rawcdn.githack.com/thephoenixofthevoid/reflect-metadata-ng/8780ddf3402e248ef4c0c663924d9eb8ef154756/diff_original_ng.html

**Reply by achrinza** [reply]
Echoing what @raymondfeng has mentioned, I think it's an interesting project that LoopBack 4 can benefit from.

I was wondering the feasibility of pushing these changes upstream, but it seems like `reflect-metadata` hasn't been updated for the past 2 years. This also means that there isn't much effort to keep `reflect-metadata-ng` in sync with upstream.

There's also not much discussion on dropping shims either: https://github.com/rbuckton/reflect-metadata/issues/73

LoopBack 4 only support LTS and Current releases of Node.js, so compatibility with v4 is not a problem.

As an interim opinion, I think it's safe to consider using `reflect-metadata-ng` as a drop-in replacement as long as the original test suite passes and is comprehensive enough.

**Reply by thephoenixofthevoid** [reply]
I would propose to simply copy the .ts file into metadata package and replace importing the original one with a new one.

Explanation:
1) This code is highly unlikely to ever change.
2) Initially I created it as a file to just drop into any project that uses decorators, leaving it runtime dependency free.
3) If author would wanted to release library without shims, it would have taken him approximately 1 hour of work.  So at least, it is not wanted.
4) I don't think they will make this change upstream: they have to stay compatible with internet explorer.
5) After all, at some point this API may be implemented natively, so maybe they simply wait for it to happen
6) I don't even have a package published, just copying it into folder feels much simpler given nor me, nor author of an original one ever will touch the implementation part of the code

**Reply by thephoenixofthevoid** [reply]
Well, my package doesn't break everything when loaded two or more times. Technically it's a breaking change. Just kidding

**Reply by achrinza** [reply]
I've dug a bit more into the repo, one concern is with the copyright; Since the project is based from https://github.com/rbuckton/reflect-metadata, we may need to retain the original Apache 2.0 license along with `AUTHORS`, `LICENSE` and `CopyrightNotice.txt` files.

Other than that, I agree with all your points above.

**Reply by thephoenixofthevoid** [reply]
There is an another way: implement missing 3 functions of that yet another implementation (it is mentioned in my README). I may implement them and cross fingers nothing else breaks. And I expect you to thoroughly inspect possible differences in behavior. If it's ok, I will fill the gap and we will have MIT licensed code after that.

**Reply by thephoenixofthevoid** [reply]
> I've dug a bit more into the repo, one concern is with the copyright; Since the project is based from https://github.com/rbuckton/reflect-metadata, we may need to retain the original Apache 2.0 license along with `AUTHORS`, `LICENSE` and `CopyrightNotice.txt` files.
> 
> Other than that, I agree with all your points above.

https://github.com/thephoenixofthevoid/reflect-metadata-ng/blob/main/TESTRUN_MIT.md
https://github.com/thephoenixofthevoid/reflect-metadata-ng/blob/main/Reflect.MIT.ts
https://github.com/thephoenixofthevoid/reflect-metadata-ng/commit/8cff50344f8cd660f7a3f2eb9f62e0b12fe71475

self-explanatory

**Reply by thephoenixofthevoid** [reply]
And finally

https://github.com/thephoenixofthevoid/reflect-metadata-ng/commit/1bd0f341864f4d2159ab0022f26275187f928cc0
https://github.com/thephoenixofthevoid/reflect-metadata-ng/commit/56e919da373e32f1527e7ac09455ed7214ec07a5

**Reply by thephoenixofthevoid** [reply]
It looks like I know cases where two packages differ: when annotation value is Falsy. But tests seem to not care about testing this edge case.
